Update documentation to include Protocol 19 features (#673)
stellar/new-docs: content/docs/glossary/transactions.mdx

Minimum sequence age is a precondition relating to time, but unlike [time bounds](#time-bounds) which express absolute times, minimum sequence age is relative to when the transaction source account's sequence number was touched.

#### Minimum Sequence Ledger Gap
When the ledger gap precondition is set, the transaction is only valid after the current network ledger number meets (or exceeds) a particular gap relative to the ledger corresponding to the account's [sequence number age](./accounts.mdx#sequence-time-and-ledger).

This is similar to the [minimum sequence age](#minimum-sequence-age), except it's expressed as a number of ledgers rather than a duration of time.

#### Extra Signers
You can require that the transaction have up to two additional signatures beyond the ones required for transaction's authorization in general. There must additional, one-off signatures corresponding to every extra signer in this list, even if the signature is not otherwise required by the sourceAccount or operations.

> The additional signers can be of any type besides a pre-authorized transaction signer, since to pre-authorize a transaction, you need to know its hash, but the hash must include the extra signers. This Catch-22 relationship means including this type of extra signer will return a [`MALFORMED` error](#result-codes).

Reviewer: Returning `MALFORMED` when trying to use pre-auth signers in extra signers didn't make it into the CAP. I discussed this with Nico and we landed on not special casing this.

Author: Is there no error, then? Just a logic error on behalf of the submitter?

Reviewer: It would end up failing with `txBAD_AUTH` because the signatures won't be satisfied.

Author: Ah yes

For some of these, it's important to understand a related concept: the last time an account consumed a sequence number. Both the ledger number and timestamp of the last time this occurred is tracked by the network, and these values enable some of the preconditions below.

#### Time Bounds
The optional UNIX timestamp (in seconds), determined by ledger time, of a lower and upper bound of when this transaction will be valid. If a transaction is submitted too early or too late, it will fail to make it into the transaction set. `maxTime` equal `0` means that it's not set. _We highly advise for all transactions to use time bounds, and many SDKs enforce their usage._ If a transaction doesn't make it into the transaction set, it is kept around in memory in order to be added to the next transaction set on a best-effort basis. Because of this behavior, we highly advise that all transactions are created with time bounds in order to invalidate transactions after a certain amount of time, especially if you plan to resubmit your transaction at a later time.

#### Ledger Bounds

> These are like [Time Bounds](#time-bounds), except they apply to ledger numbers. With them set, a transaction will only be valid for ledger numbers that fall into the range you set. The lower bound is inclusive while the upper bound is not. If you set the upper bound to zero, this indicates that there is no upper bound.

Reviewer 2: >The lower bound is inclusive while the upper bound is not.

This is incorrect. The upper bound is inclusive.

See https://github.com/stellar/stellar-core/blob/da846c923bbd989fe716c9a3eb32e04ef2ae37db/src/transactions/TransactionFrame.cpp#L491-L496.

cc @sisuresh for confirmation

Author: Then the [CAP-21 comments](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0021.md#xdr-diff) are incorrect:

```diff
+    // Transaction only valid for ledger numbers n such that
+    // minLedger <= n < maxLedger (if maxLedger == 0, then
+    // only minLedger is checked)
+    LedgerBounds *ledgerBounds;
```

#### Minimum Sequence Number

> If this precondition is set, the transaction will only be valid when its source account's sequence number (call it `S`) is large enough. Specifically, it's valid when `S` satisfies `minSeqNum <= S < tx.seqNum`. If this precondition is omitted, the default behavior applies: the transaction's sequence number must be exactly one greater than the account's sequence number.

Reviewer 2: I think we should avoid the "this precondition". It's helpful to imagine that the headings are gone and the paragraphs should speak for themselves.
```suggestion
If a minimum sequence number is set, the transaction will only be valid when its source account's sequence number (call it `S`) is large enough. Specifically, it's valid when `S` satisfies `minSeqNum <= S < tx.seqNum`. If this precondition is omitted, the default behavior applies: the transaction's sequence number must be exactly one greater than the account's sequence number.
```

Author: But then what's the point of the headings? Especially since it's the first paragraph following it. Seems redundant.

Reviewer 3: +1 on the suggestion.  Not only is it better for cut-and-paste, but it will also make the copy parallel the sections below, which start with "when the sequence age precondition is set..." and "when the ledger gap precondition is set..."

I think the titles serve a separate purpose: they're signposts, and they show up on the nav menu so people can find and jump to a given section.

> Note that after a transaction is executed, the account will always raise its sequence number to the transaction's sequence number.

Reviewer 2: I think this is subtle, but I think technically we describe this as setting, not raising, because it doesn't work like bumpseq's where it only changes it if it is lower. This is not particularly meaningful because of the constraint described in the prior paragraph we know it nevers goes backwards, but mostly we describe it as 'set' in our own conversations.
```suggestion
Note that after a transaction is executed, the account will always set its sequence number to the transaction's sequence number.
```

#### Minimum Sequence Age

> With this precondition, the transaction is only valid after an account's last sequence number change exceeds a particular duration (expressed in seconds).

Reviewer 2: This is subtly incorrect.
```suggestion
Minimum sequence age makes a transaction valid only if a duration of time (specified in seconds) has passed since the last time that the sequence number for a source account has been operated on, either by a BumpSequence operation, or a Transaction. Note that a BumpSequence resets this period regardless of whether the sequence number of the source account is changed.
```

> With a minimum [time bound](#time-bounds), the transaction becomes valid after a particular amount of time passes unconditionally. With this precondition, the passage of time applies only to the transaction's source account: you can allow validity after an *account* has waited long enough since its last sequence number change.

Reviewer 2: This paragraph is confusing, and it feels unrelated to this section, or out of place. Is it a copy-paste mistake?

Author: What's confusing about it? It's relating the concept of timebound-based validity to "account did a thing"-based validity.

Reviewer 2: The fact it leads with discussing time bounds. Maybe rewording it like this would help:

```suggestion
Minimum sequence age is a precondition relating to time, but unlike [time bound](#time-bounds) which is an absolute time, minimum sequence age is relative to the transaction source account's sequence age field.
```

Author: Ooh, very nice reword.

#### Minimum Sequence Ledger Gap

> With this precondition, the transaction is only valid after the current network ledger number meets (or exceeds) a particular gap relative to the ledger corresponding to the source account's last sequence number change.

Reviewer 2: Same as above:
```suggestion
Minimum sequence ledger gap makes a transaction valid only if a duration of ledgers has passed since the last ledger that the sequence number for a source account has been operated on, either by a BumpSequence operation, or a Transaction. Note that a BumpSequence resets this period regardless of whether the sequence number of the source account is changed.
```

Author: I consolidated this into parent paragraph - _Validity Conditions_ - and clarify/introduce/define what it means for an account sequence number to be touched (I also think a term like `refreshed` could work).

This is similar to the [minimum sequence age](#minimum-sequence-age), except it's expressed as a quanity of ledgers rather than a duration of time.

#### Extra Signers

> You can require that the transaction have up to two additional signatures beyond the ones required for transaction's authorization in general. There must additional, one-off signatures corresponding to every extra signer in this list, even if the signature is not otherwise required by the sourceAccount or operations.

Reviewer 2: Subtle changes regarding signers vs signatures:
```suggestion
You can require that the transaction have up to two additional signers beyond the ones required for transaction's authorization in general. For each extra signer there must be a signature corresponding to the signer, even if the signature would not otherwise be required by the transaction's source account or operations.
```

Author: By adding extra signers, you require extra signatures on the transaction. To say you can require extra signers doesn't really make sense.

Perhaps this is a better reword?
```suggestion
A transaction can specify up to two extra signers as a precondition, meaning it must have corresponding additional signatures, even if those signatures would not otherwise be required to authorize the transaction (i.e. for its source account or operations).
```

Reviewer 2: My edit says there has to be a signature corresponding to the signer, not that it requires extra signers. An extra signer may require the same signature as an operation already does, so strictly speaking an extra signer doesn't require an extra signature, or even any additional signatures, there just has to be corresponding signatures to signers.

Your edit is fine too, if you remove the `additional` word:
```suggestion
A transaction can specify up to two extra signers as a precondition, meaning it must have signatures that correspond to those extra signers, even if those signatures would not otherwise be required to authorize the transaction (i.e. for its source account or operations).
```

Author: It does say that, though, doesn't it? :smile: 

> You can require that the transaction have up to two additional signers 

Good point regarding redundant signatures, though.

  - The appropriate network passphrase was part of the transaction hash that was signed by each of the signers. See [Network Passphrases](./network-passphrase.mdx) for more.
  - The combined weight of all signatures for the source account _of the transaction_ meets the low threshold for the source account. This is necessary in order for fees to be taken and the sequence number to be incremented later in the transaction lifecycle.
- **Memo** — The memo type must be a valid type, and the memo itself must be adhere to the formatting of the memo type.
- **Preconditions** — refer to the [Validity Conditions](#validity-conditions) described above for thorough explanations of each of these. For each of these that are set, the transaction will be considered invalid unless the transaction is submitted:
  * **Time Bounds** — within the set time bounds of the transaction
  * **Ledger Bounds** — within the set ledger bounds of the transaction

>   * **Minimum Sequence Number** — by a source account whose sequence number is larger than this value

Reviewer 2: Correction:
```suggestion
  * **Minimum Sequence Number** — by a source account whose sequence number is larger or equal to this value
```

Author: With that, "greater than" is more conventional than "larger"

```suggestion
  * **Minimum Sequence Number** — by a source account whose sequence number is greater than or equal to this value
```

>   * **Minimum Sequence Age** — by a source account who last changed its sequence number at least "x" long ago

Reviewer 2: This is incorrect, it doesnt' have to have changed, but it is also hard to discuss it here, so I think we should just link to teh larger section.
```suggestion
  * **Minimum Sequence Age** — see [Minimum Sequence Age](#minimum-sequence-age).
```

Author: Yeah seems good, amended.

>   * **Minimum Sequence Ledger Gap** — by a source account whose sequence number was last changed in a ledger at least "x" ledgers from the current ledger

Reviewer 2: Same as above:
```suggestion
  * **Minimum Sequence Ledger Gap** — see [Minimum Sequence Ledger Gap](#minimum-sequence-ledger-gap).
```
